web/src/upload.ts

```diff
@@ -464,22 +485,34 @@ export function setup_upload(config: Config): Uppy<Meta, TusBody> {
 
     uppy.on("upload-success", (file, response) => {
         assert(file !== undefined);
-        let upload_response;
+        if (response.status !== 200) {
+            blueslip.warn("Tus server returned an error, expected a 200 OK response code.", {
+                response,
+            });
+            return;
+        }
+
         try {
-            upload_response = zulip_upload_response_schema.parse(
+            const upload_response = zulip_upload_response_schema.parse(
                 JSON.parse(response.body!.xhr.responseText),
             );
-        } catch {
-            blueslip.warn("Invalid JSON response from tus server", {
-                body: response.body!.xhr.responseText,
+            // Change filename to the one provided by the server.
+            uppy.setFileState(file.id, {
+                name: upload_response.filename,
+                uploadURL: upload_response.url,
             });
-            return;
+            file = uppy.getFile(file.id);
+        } catch {
+            // We do not receive a response body if the file has already
+            // been uploaded. Instead a HEAD request is sent and we
+            // get the uploadURL back. In case of an already uploaded
+            // file, we would have updated the filename and uploadURL
+            // accordingly in the onBeforeFileAdded hook and won't need
+            // to do anything in this catch clause.
```

Comment:
Hm.  Blindly swallowing all errors, which is what an empty `catch` block does, is a pretty bad code smell -- we know nothing about what's actually happening here, besides that it's a 200 which doesn't meet the schema.  It may not be JSON.  It may be a tusd upgrade which changes the schema somehow.

We should do some checking that we're in the semi-happy path of having just sent a HEAD and gotten an uploadURL back, instead of assuming that every error is that case.  I can imagine various situations (e.g. a captive portal interstitial) which result in getting back a 200 with abnormal content, which we'd want to raise an error and not attempt to blindly continue.